### DISCUSSION TITLE
Initial Zephyr RTOS support

### DISCUSSION THREAD

**Discussion by henrikbrixandersen** [stnolting/neorv32#172] [Show and tell]
I have submitted a PR for the Zephyr RTOS for adding initial support for running on the NEORV32 SoC: https://github.com/zephyrproject-rtos/zephyr/pull/39162

Once the initial support has been merged, I will submit individual PRs for adding support for the other NEORV32 peripherals.

**Comment by stnolting** [comment]
This is really awesome!
I always wanted to check out Zephyr on the NEORV32, but I was overwhelmed by the porting complexity... 😅 
I am really looking forward to this - your commits at Zephyr are just like a porting guide.

Thank you so much for your great work! :+1:

**Comment by stnolting** [comment]
@henrikbrixandersen 

Your port is upstream! :rocket:
This is really really great - thanks again for porting! :+1:

Right now I am trying to test the basic "hello world" example (https://docs.zephyrproject.org/latest/boards/riscv/neorv32/doc/index.html). Setting up the build system is quite easy and _west_ is incredible complex but very handy.

My first attempt to generate an executable failed, because the NEORV32 executable image generator was not in `PATH`. Maybe it would be good to add that to the documentation, too.

```
-- The neorv32 image_gen utility was not found, neorv32 image files cannot be generated
```

 However, after adding that everything compiled without problems (tested with the latest hardware version [`1.6.2.1`](https://github.com/stnolting/neorv32/blob/master/CHANGELOG.md)):

```
CMD:> u
Awaiting neorv32_exe.bin... OK
CMD:> e
Booting...

*** Booting Zephyr OS build v2.7.99-547-g02fceed732ab  ***
Hello World! neorv32
```

**:tada: Zephyr OS on the NEORV32! :tada: :smile:**

-------

Just a small question regarding the ELF generated by the Zephyr build system:

I tried to convert `zephyr.elf` into an executable "by hand", so using the default NEORV32 makefile. This does not work because the NEORV32 linker script puts all the interesting stuff in sections like `.text` (names starting with a **dot**) while the Zephyr linker script puts everything in sections like `text` (NO dot). Is this intentional?

**Reply by henrikbrixandersen** [reply]
Yes, it didn't take long. I will start submitting my NEORV32 peripheral drivers to Zephyr (along with support for the UART FIFO, thank you for that addition) as PRs against Zephyr.

I will submit a clarification on how to get CMake to detect the NEORV32 `image_gen` utility to the documentation as well, appreciate the input.

The ELF section in Zephyr are named according to the conventions used throughout the Zephyr RTOS, which does not use the dot-name notation.

**Reply by stnolting** [reply]
> Yes, it didn't take long. I will start submitting my NEORV32 peripheral drivers to Zephyr (along with support for the UART FIFO, thank you for that addition) as PRs against Zephyr.

:+1:

> I will submit a clarification on how to get CMake to detect the NEORV32 image_gen utility to the documentation as well, appreciate the input.

:+1: :+1:

> The ELF section in Zephyr are named according to the conventions used throughout the Zephyr RTOS, which does not use the dot-name notation.

Thanks for clearing!

Just another thing regarding the frequency base:
The hello-world example uses clock information from the SYSINFO module to configure the baud rate. The blink example uses the frequency specified in the device tree (at least this is what I think...). This is not consistent but I think it is a real nice "fail-safe" feature - at least for me: I could run all the demos on my 24Mhz setup without changing a single line of code.

**Reply by henrikbrixandersen** [reply]
> I will submit a clarification on how to get CMake to detect the NEORV32 `image_gen` utility to the documentation as well, appreciate the input.

PR submitted as https://github.com/zephyrproject-rtos/zephyr/pull/39772. Hopefully this will make it more clear.

**Reply by henrikbrixandersen** [reply]
> Just another thing regarding the frequency base:
> The hello-world example uses clock information from the SYSINFO module to configure the baud rate. The blink example uses the frequency specified in the device tree (at least this is what I think...). This is not consistent but I think it is a real nice "fail-safe" feature - at least for me: I could run all the demos on my 24Mhz setup without changing a single line of code.

Currently, the Zephyr MTIME driver depends on `CONFIG_SYS_CLOCK_TICKS_PER_SEC` (which is derived from the devicetree on NEORV32) for determining the timer frequency. A few other Zephyr system timer drivers support `CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME`, but this is not (yet) supported by the MTIME driver. I have added it to my TODO.

**Reply by robinmeister** [reply]
Hey i am trying to use the Zephyr RTOS on the neorv32 soft-core and so far everything runs fine except the k_sleep() Function. Every time i try to use the function the program doesn't continue executing. The documentation says that the interrupt controller and system timer are supported features so I assumed the function would just work. Do you have any ideas what is going wrong there? I use Zephyr sdk 0.15.2.

**Reply by henrikbrixandersen** [reply]
> Hey i am trying to use the Zephyr RTOS on the neorv32 soft-core and so far everything runs fine except the k_sleep() Function. Every time i try to use the function the program doesn't continue executing. The documentation says that the interrupt controller and system timer are supported features so I assumed the function would just work. Do you have any ideas what is going wrong there? I use Zephyr sdk 0.15.2.  

Which version of Zephyr (not the SDK, Zephyr itself)? Which version of neorv32?

**Reply by robinmeister** [reply]
Zephyr OS Version: 3.1.99
GNU RISCV toolchain version 11.1.0
NEORV32 reposiory release version 1.8.0
The Setup for all of this is done through a install.sh file from the following project: 
https://github.com/Lien182/ReconOS_Zephyr_RISCV/tree/master

**Reply by stnolting** [reply]
Sorry for the late reply... 🙈

There is a (work-in-progress) PR to update the Zephyr port according to the processor's new memory layout: https://github.com/zephyrproject-rtos/zephyr/pull/65627

**Reply by kiriane-c** [reply]
> @henrikbrixandersen
> 
> Your port is upstream! 🚀 This is really really great - thanks again for porting! 👍
> 
> Right now I am trying to test the basic "hello world" example (https://docs.zephyrproject.org/latest/boards/riscv/neorv32/doc/index.html). Setting up the build system is quite easy and _west_ is incredible complex but very handy.
> 
> My first attempt to generate an executable failed, because the NEORV32 executable image generator was not in `PATH`. Maybe it would be good to add that to the documentation, too.
> 
> ```
> -- The neorv32 image_gen utility was not found, neorv32 image files cannot be generated
> ```
> 
> However, after adding that everything compiled without problems (tested with the latest hardware version [`1.6.2.1`](https://github.com/stnolting/neorv32/blob/master/CHANGELOG.md)):
> 
> ```
> CMD:> u
> Awaiting neorv32_exe.bin... OK
> CMD:> e
> Booting...
> 
> *** Booting Zephyr OS build v2.7.99-547-g02fceed732ab  ***
> Hello World! neorv32
> ```
> 
> **🎉 Zephyr OS on the NEORV32! 🎉 😄**
> 
> Just a small question regarding the ELF generated by the Zephyr build system:
> 
> I tried to convert `zephyr.elf` into an executable "by hand", so using the default NEORV32 makefile. This does not work because the NEORV32 linker script puts all the interesting stuff in sections like `.text` (names starting with a **dot**) while the Zephyr linker script puts everything in sections like `text` (NO dot). Is this intentional?

Hey @stnolting,

I also tried to test the basic "hello_world" example on the NeoRV32 via the bootloader. I included the path for image_gen and I got the .bin file that I uploaded. But when I try to execute, I'm stuck in the boot as such :

```
CMD:> u
Awaiting neorv32_exe.bin... OK
CMD:> e
Booting 0x00000000...
```
Could you please tell me what I've probably done wrong ?
Thanks

**Reply by henrikbrixandersen** [reply]
> I also tried to test the basic "hello_world" example on the NeoRV32 via the bootloader. I included the path for image_gen and I got the .bin file that I uploaded. But when I try to execute, I'm stuck in the boot as such :
> 
> CMD:> u
> Awaiting neorv32_exe.bin... OK
> CMD:> e
> Booting 0x00000000...
> Could you please tell me what I've probably done wrong ?

Which version of NEORV32? Which version of Zephyr?

Did you try attaching a debugger to see where it gets stuck?

**Reply by kiriane-c** [reply]
I am using Zephyr 3.6.0 and NEORV32 RISC-V Processor (version 0x01090803). 

I also tried to debug but I get this and I don't really know what it means.

```
-- west debug: rebuilding
ninja: no work to do.
-- west debug: using runner openocd
-- runners.openocd: OpenOCD GDB server running on port 3333; no thread info available
Open On-Chip Debugger 0.11.0+dev-00728-gb6f95a16c (2024-05-29-12:13)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
1
adapter speed: 2000 kHz

Error: no device found
Error: unable to open ftdi device with vid 0403, pid 6010, description 'Flyswatter', serial '*' at bus location '*'

GNU gdb (Zephyr SDK 0.16.8) 12.1
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "--host=x86_64-build_pc-linux-gnu --target=riscv64-zephyr-elf".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://github.com/zephyrproject-rtos/sdk-ng/issues>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /zephyrproject/zephyr/build/zephyr/zephyr.elf...
:3333: Connexion terminée par expiration du délai d'attente.
You can't do that when your target is 'exec'
```

**Reply by stnolting** [reply]
> Could you please tell me what I've probably done wrong ?

That's hard to say. What processor configuration are you using?
* are you using internal IMEM & DMEM?
* what are the sizes of the memories (e.g. IMEM + DMEM)? Is this sync with the linker script / Zephyr build?
* have you enabled the MTIME machine time?
* what CPU ISA extensions are you using? Is this sync with your Zephyr build?

> I also tried to debug but I get this and I don't really know what it means.

Seems like there is no JTAG adapter available. What kind of adapter are you using? Is it a (simple) FTDI-based one?

**Reply by kiriane-c** [reply]
I'm using internal IMEM and DMEM with sizes 64kB and 32kB respectively. MTIME and the CPU extensions C, M and Zicsr are enabled. 

I think the problem comes from the FPGA board I'm using; Zynq UltraScale+ MPSoC ZCU104. This is because when trying to access JTAG boot mode, I get errors on Vivado regarding Power Status and POR_B signal.

**Reply by stnolting** [reply]
The processor configuration seems fine (assuming that you have also enabled the UART).

Why do you think there is a FPGA issue? You can see the bootloader console, right?

**Reply by henrikbrixandersen** [reply]
> I am using Zephyr 3.6.0 and NEORV32 RISC-V Processor (version 0x01090803).

Zephyr v3.6.0 supports NEORV32 v1.8.6.

**Reply by kiriane-c** [reply]
@stnolting, I think there is a hardware issue when using the JTAG mode of my FPGA. Vivado doesn't detect the FPGA in that mode and signals warnings/errors on the POR_B signal.

**Reply by stnolting** [reply]
Oh, that's not good... But maybe this is just some boot configuration option issue?! Does your board provide some switches/jumper to select the FPGA boot mode? Perhaps something is configured incorrectly there.

**Comment by henrikbrixandersen** [comment]
Zephyr TRNG support PR: https://github.com/zephyrproject-rtos/zephyr/pull/39527

**Comment by robhancocksed** [comment]
FYI, I've got a PR to update the Zephyr port to the 1.9.8 core revision and add in a few more drivers like XIRQ and the watchdog timer: https://github.com/zephyrproject-rtos/zephyr/pull/71294

The recent neorv32 core changes in interrupt handling make it a lot easier to support interrupts in a general fashion across drivers - previously the way in which FIRQ interrupts needed to be cleared was fairly inconsistent.

There's some discussion about whether it's necessary to maintain compatibility with previous neorv32 revisions. It seems difficult to do so given that the differences go beyond what would end up in the device tree files and extend to things like the bit assignments in SYSINFO registers etc.

Part of the answer might depend on whether there are known "breaking" changes coming in core parts of the neorv32 core which might result in even more versions that need to be supported?

**Comment by stnolting** [comment]
Hey @robhancocksed!

> I've got a PR to update the Zephyr port to the 1.9.8 core revision and add in a few more drivers like XIRQ and the watchdog timer: https://github.com/zephyrproject-rtos/zephyr/pull/71294

This looks really great! I have already received a few requests asking when there will be an updated version of the NEORV32 port. So, thank you very much for all your work!

> The recent neorv32 core changes in interrupt handling make it a lot easier to support interrupts in a general fashion across drivers - previously the way in which FIRQ interrupts needed to be cleared was fairly inconsistent.

Right, the FIRQ handling was quite... horrible... 😅 I'm really glad that we finally reworked that! The FIRQ handling is not only much more consistent among each other, but also in relation to the other standard RISC-V interrupts.

> There's some discussion about whether it's necessary to maintain compatibility with previous neorv32 revisions. It seems difficult to do so given that the differences go beyond what would end up in the device tree files and extend to things like the bit assignments in SYSINFO registers etc.

That's indeed a problem.

In my humble opinion, only the latest/current version should be supported. Anything else would be far too much work in the end... 🙈

> Part of the answer might depend on whether there are known "breaking" changes coming in core parts of the neorv32 core which might result in even more versions that need to be supported?

Furthermore, this project is under constant development. Breaking changes are not planned, but they happen from time to time, especially when we discover a severe bug. And isn't such a bug-fix ultimately a "breaking" change? 🤔

**Comment by koviadjevi** [comment]
Hello @all, i am currently trying to run the basic Zephyr  Helloworld application on a digilent Genesys2 board. 
Following the Zephyr documentation's instructions for the Neorv32, i succeed to use **west** for the build, but i am currently stuck at the step of flashing, since i am using JTAG 
As required by the documentation i a enabled the **OCD_EN** param et the **bootmode is 0**. But when running the **west flash** it results in the following error message 

###################################################
Open On-Chip Debugger 0.11.0+dev-00728-gb6f95a16c-dirty (2024-10-20-01:26)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
1
Info : clock speed 1000 kHz
Error: JTAG scan chain interrogation failed: all zeroes
Error: Check JTAG interface, timings, target power, etc.
Error: Trying to use configured scan chain anyway...
Error: neorv32.cpu: IR capture error; saw 0x00 not 0x01
Warn : Bypassing JTAG setup events due to errors
Error: dtmcontrol is 0. Check JTAG connectivity/board power.
Warn : target neorv32.cpu.0 examination failed
Info : starting gdb server for neorv32.cpu.0 on 3333
Info : Listening on port 3333 for gdb connections
    TargetName         Type       Endian TapName            State       
--  ------------------ ---------- ------ ------------------ ------------
 0* neorv32.cpu.0      riscv      little neorv32.cpu        unknown

Error: Target not examined yet

FATAL ERROR: command exited with status 1: ..../zephyr-sdk-0.17.0/sysroots/x86_64-pokysdk-linux/usr/bin/openocd -s ...../zephyr/zephyrproject/zephyr/boards/others/neorv32/support -s .....zephyr-sdk-0.17.0/sysroots/x86_64-pokysdk-linux/usr/share/openocd/scripts -f ...../zephyr/zephyrproject/zephyr/boards/others/neorv32/support/openocd.cfg '-c init' '-c targets' -c halt -c 'load_image ....../zephyr/zephyrproject/zephyr/build/zephyr/zephyr.elf' -c 'resume 0x0000000000000000' -c shutdown
###################################################

Do  you guys have an idea of what i may be doing wrong and how i can fix it please ?

**Reply by henrikbrixandersen** [reply]
> Error: Check JTAG interface, timings, target power, etc.

Did you check this? All zeros typically means you've wired up your JTAG wrong (or that no target is on the other end of the JTAG lines). 

As listed in the documentation, the provided JTAG setup is just an example. You may need to adjust it to fit your particular JTAG adapter.

**Reply by koviadjevi** [reply]
> Did you check this? All zeros typically means you've wired up your JTAG wrong (or that no target is on the other end of the JTAG lines).

Thank you @henrikbrixandersen  for your quick reply. 
Yes as for checking the wiring up of the JTAG i did check and the board (Genesys2) is ON, JTAG is working (since i succeed to upload the NEORV32 design bitstream into the board FPGA).
 If there is a need to check something else, please dont hesitate to let me know, thanks again for the quick reply.

**Reply by koviadjevi** [reply]
> As listed in the documentation, the provided JTAG setup is just an example. You may need to adjust it to fit your particular JTAG adapter.

I think that may be what i am missing, how can that be done, wich of the files of config i need to change, pls?

**Reply by henrikbrixandersen** [reply]
> Yes as for checking the wiring up of the JTAG i did check and the board (Genesys2) is ON, JTAG is working (since i succeed to upload the NEORV32 design bitstream into the board FPGA).

My guess is that you do not have the NEORV32 JTAG signals wired up to the same pins as the FPGA JTAG interface? There are two JTAG endpoints in play here.

**Reply by robhancocksed** [reply]
> > Did you check this? All zeros typically means you've wired up your JTAG wrong (or that no target is on the other end of the JTAG lines).
> 
> Thank you @henrikbrixandersen for your quick reply. Yes as for checking the wiring up of the JTAG i did check and the board (Genesys2) is ON, JTAG is working (since i succeed to upload the NEORV32 design bitstream into the board FPGA). If there is a need to check something else, please dont hesitate to let me know, thanks again for the quick reply.

The NEORV32 CPU has its own JTAG endpoint, separate from the one that the Xilinx FPGA uses for its configuration. You'll need to route those JTAG lines from the core to some physical pins on the board that you can connect your JTAG adapter to.

It would be nice if there was a way to have this piggy-back on the FPGA's internal JTAG port (like you can do with the Xilinx MicroBlaze cores), but I'm not sure there is a simple way to do this right now.

**Reply by stnolting** [reply]
Right. You'll need to connect the processor's JTAG lines to an external adapter:

https://github.com/stnolting/neorv32/blob/3359f5fa40b98e76c396f47f8551a2d1ea084887/rtl/core/neorv32_top.vhd#L152-L156

You can find more information in the according section of the user guide: https://stnolting.github.io/neorv32/ug/#_debugging_using_the_on_chip_debugger

-------------

Btw, it is also possible to use the JTAG interface, which is used for the configuration of the FPGA, for the user logic (i.e. the processor). But for this you have to implement special primitives with some glue logic. You also need a specific configuration of openOCD.

**Reply by koviadjevi** [reply]
> You'll need to connect the processor's JTAG lines to an external adapter:

Thank you for the reply, 
i did try your suggestions then regenerate the bitstream and rebuid zephyr with west and all... but still not working, i am still getting the same exact error message. Here is what i did: 
Since there is no external JTAG pin from the board what is suggested is to use a **BSAN** module from xiling then connect its signals. 
So i did the following:
- for the neorv32 instance
~~~

....
....
signal tck_i, tdi_i, tdo_o, tms_i : std_logic;
......
.....

neorv32_top_inst: neorv32_top
  generic map (
    -- Clocking --
    CLOCK_FREQUENCY   => CLOCK_FREQUENCY,   -- clock frequency of clk_i in Hz
    OCD_EN              => true,                     -- implement on-chip debugger
    -- Boot Configuration --
    BOOT_MODE_SELECT  => 0,                 -- boot via internal bootloader
    -- RISC-V CPU Extensions --
    RISCV_ISA_C       => true,              -- implement compressed extension?
    RISCV_ISA_M       => true,              -- implement mul/div extension?
    RISCV_ISA_Zicntr  => true,              -- implement base counters?
    -- Internal Instruction memory --
    MEM_INT_IMEM_EN   => true,              -- implement processor-internal instruction memory
    MEM_INT_IMEM_SIZE => MEM_INT_IMEM_SIZE, -- size of processor-internal instruction memory in bytes
    -- Internal Data memory --
    MEM_INT_DMEM_EN   => true,              -- implement processor-internal data memory
    MEM_INT_DMEM_SIZE => MEM_INT_DMEM_SIZE, -- size of processor-internal data memory in bytes
    -- Processor peripherals --
    IO_GPIO_NUM       => 8,                 -- number of GPIO input/output pairs (0..32)
    IO_CLINT_EN       => true,              -- implement core local interruptor (CLINT)?
    IO_UART0_EN       => true               -- implement primary universal asynchronous receiver/transmitter (UART0)?
  )
  port map (
    -- Global control --
    clk_i       => s_clk,        -- global clock, rising edge
    rstn_i      => s_rst_n,       -- global reset, low-active, async
    -- GPIO (available if IO_GPIO_NUM > 0) --
    gpio_o      => con_gpio_out, -- parallel output
    -- primary UART0 (available if IO_UART0_EN = true) --
    uart0_txd_o => uart0_txd_o,  -- UART0 send data
    uart0_rxd_i => uart0_rxd_i,   -- UART0 receive data
    jtag_tck_i          => tck_i,
    jtag_tdi_i          => tdi_i,
    jtag_tdo_o          => tdo_o,
    jtag_tms_i          => tms_i
  );
~~~ 

for the BSCAN instance 
~~~
jtag_bscan_instance: 
component BSCANE2
port map (
      tck => tck_i,
      tdi => tdi_i,
      tms => tms_i,
      tdo => tdo_o
   );
~~~
Is there anything that i may be missing in my code please ?

**Reply by stnolting** [reply]
Unfortunately, you cannot use the BSCAN primitives as a "normal" JTAG port because they are not directly connected to the JTAG pins of the FPGA. Instead, the BSCAN primitives use specific JTAG instructions to build something like a "tunnel" that can be used to issue custom JTAG instructions.

You need some glue logic that converts the BSCAN operations into debug module interface transactions (and not the JTAG transactions of the actual debug transfer module).

Basically, you need something like this: https://github.com/eugene-tarassov/vivado-risc-v/blob/master/vhdl-wrapper/src/net/largest/riscv/vhdl/jtag.vhdl

This is pretty complicated. I recommend mapping the processor's JTAG signals to some headers on your board and connecting a simple USB-JTAG bridge (e.g. an FTDI break-out board).

**Reply by koviadjevi** [reply]
@stnolting i am not sure i get it correctly, will that be possible to give a code example on how that can be done please?

**Reply by stnolting** [reply]
I am not familiar with the Genesys 2 board. However, I am using the Nexys Video board which provides a secondary USB/FTDI port that can be used as _user_ JTAG port for the processor's on-chip debugger.

Here is a cut-out from my (Nexys Video) XDC file that connects the processor's JTAG lines to the secondary USB-JTAG bridge:

```xdc
set_property -dict { PACKAGE_PIN U20   IOSTANDARD LVCMOS33 } [get_ports { jtag_tck_i }]; #IO_L11P_T1_SRCC_14 Sch=prog_d0/sck
set_property -dict { PACKAGE_PIN P14   IOSTANDARD LVCMOS33 } [get_ports { jtag_tdi_i }]; #IO_L19P_T3_A10_D26_14 Sch=prog_d1/mosi
set_property -dict { PACKAGE_PIN P15   IOSTANDARD LVCMOS33 } [get_ports { jtag_tdo_o }]; #IO_L22P_T3_A05_D21_14 Sch=prog_d2/miso
set_property -dict { PACKAGE_PIN U17   IOSTANDARD LVCMOS33 } [get_ports { jtag_tms_i }]; #IO_L18P_T2_A12_D28_14 Sch=prog_d3/s
```

Looking at the [XDC file of Genesys 2](https://github.com/Digilent/digilent-xdc/blob/00a3404901f35aa9567b01ecb3f2c233b6efe9f4/Genesys-2-Master.xdc#L373-L376) it seems like this is also possible for that board:

```xdc
#set_property -dict { PACKAGE_PIN AD27  IOSTANDARD LVCMOS33 } [get_ports { PROG_D[0] }]; #IO_L11P_T1_SRCC_13 Sch=prog_d0/sck
#set_property -dict { PACKAGE_PIN W27   IOSTANDARD LVCMOS33 } [get_ports { PROG_D[1] }]; #IO_L2P_T0_13 Sch=prog_d1/mosi
#set_property -dict { PACKAGE_PIN W28   IOSTANDARD LVCMOS33 } [get_ports { PROG_D[2] }]; #IO_L2N_T0_13 Sch=prog_d2/miso
#set_property -dict { PACKAGE_PIN W29   IOSTANDARD LVCMOS33 } [get_ports { PROG_D[3] }]; #IO_L4P_T0_13 Sch=prog_d3/ss
```

**Reply by koviadjevi** [reply]
Thank you very much @stnolting !!!, it works like a charm after adapting a bit your suggestions! thanks again.

**Reply by stnolting** [reply]
That's awesome! 🚀
